src/QsCompiler/Tests.Compiler/CallGraphTests.fs

```diff
@@ -675,14 +719,193 @@ type CallGraphTests (output:ITestOutputHelper) =
+
+        // The generalized methods of asserting dependencies assumes Body nodes, but
+        // this relationship is between a Body and an Adjoint specializations, so we
+        // will check manually.
+        let mainNode =
+            CallGraphNode({ Namespace = NonNullable<_>.New Signatures.TypeParameterResolutionNS; Name = NonNullable<_>.New "Main" },
+                QsSpecializationKind.QsBody, QsNullable<ImmutableArray<ResolvedType>>.Null)
+
+        let adjFooNode =
+            CallGraphNode({ Namespace = NonNullable<_>.New Signatures.TypeParameterResolutionNS; Name = NonNullable<_>.New "Foo" },
+                QsSpecializationKind.QsAdjoint, QsNullable<ImmutableArray<ResolvedType>>.Null)
+
+        let mainDependencies = graph.GetDirectDependencies mainNode
+        Assert.True(mainDependencies.Contains(adjFooNode) && mainDependencies.[adjFooNode].Any(),
+            sprintf "Expected %s to take dependency on %s." "Main" "Adjoint Foo")
+
+        [
+            "Foo", [
+                "Bar"
+            ]
+        ]
+        |> List.map (fun x -> AssertExpectedDirectDependencies (fst x) (snd x) graph)
+        |> ignore
+
+    [<Fact>]
+    [<Trait("Category","Populate Call Graph")>]
+    member this.``Not Called With Entry Point`` () =
+        let graph = CompileTypeParameterResolutionTestWithExe 14
+
+        [
+            "Main", [
+                "Foo"
+            ]
+            "Foo", [ ]
+        ]
+        |> List.map (fun x -> AssertExpectedDirectDependencies (fst x) (snd x) graph)
+        |> ignore
+
+        [
+            "NotCalled"
+        ]
+        |> List.map (AssertNotInGraph graph)
+        |> ignore
+
+    [<Fact>]
+    [<Trait("Category","Populate Call Graph")>]
+    member this.``Not Called Without Entry Point`` () =
+        let graph = CompileTypeParameterResolutionTest 15
+
+        [
+            "Main", [
+                "Foo"
+            ]
+            "Foo", [ ]
+        ]
+        |> List.map (fun x -> AssertExpectedDirectDependencies (fst x) (snd x) graph)
+        |> ignore
+
+        [
+            "NotCalled"
+        ]
+        |> List.map (AssertNotInGraph graph)
+        |> ignore
+
+    [<Fact>]
+    [<Trait("Category","Populate Call Graph")>]
+    member this.``Unrelated Without Entry Point`` () =
+        let graph = CompileTypeParameterResolutionTest 16
+
+        [
+            "Main", [
+                "Foo"
+            ]
+            "Foo", [ ]
+            "Bar", [
+                "Baz"
+            ]
+            "Baz", [ ]
+        ]
+        |> List.map (fun x -> AssertExpectedDirectDependencies (fst x) (snd x) graph)
+        |> ignore
+
+    [<Fact>]
+    [<Trait("Category","Populate Call Graph")>]
+    member this.``Entry Point No Descendants`` () =
+        let graph = CompileTypeParameterResolutionTestWithExe 17
+        Assert.True(graph.Count = 0, "Expected call graph to be empty.")
```

Comment:
Why is this an empty graph instead of a graph with one node?

Reply:
Call graphs only contain relationships between callables. If a callable is not referenced and is not being referenced anywhere, it can't be in the graph. Here the entry point does not have any relationships, and so cannot be in the graph.

Reply:
Based on our conversation, it looks like an ambiguity in the design doc, so I'll move this comment to the design doc PR.